src/Cli/Microsoft.DotNet.FileBasedPrograms/FileLevelDirectiveHelpers.cs

```diff
@@ -468,21 +518,58 @@ public sealed class Project(in ParseInfo info) : Named(info)
                 return context.Diagnostics.AddError<Project?>(context.SourceFile, context.Info.Span, string.Format(FileBasedProgramsResources.MissingDirectiveName, directiveKind));
             }
 
+            return new Project(context.Info, directiveText);
+        }
+
+        public enum NameKind
+        {
+            /// <summary>
+            /// Change <see cref="Named.Name"/> and <see cref="ExpandedName"/>.
+            /// </summary>
+            Expanded = 1,
+
+            /// <summary>
+            /// Change <see cref="Named.Name"/> and <see cref="ResolvedName"/>.
+            /// </summary>
+            Resolved = 2,
+
+            /// <summary>
+            /// Change only <see cref="Named.Name"/>.
+            /// </summary>
+            Final = 3,
+        }
+
+        public Project WithName(string name, NameKind kind)
+        {
+            return new Project(Info, name)
+            {
+                OriginalName = OriginalName,
+                ExpandedName = kind == NameKind.Expanded ? name : ExpandedName,
+                ResolvedName = kind == NameKind.Resolved ? name : ResolvedName,
+            };
+        }
+
+        /// <summary>
+        /// If the directive points to a directory, returns a new directive pointing to the corresponding project file.
+        /// </summary>
+        public Project ResolveProjectPath(SourceFile sourceFile, DiagnosticBag diagnostics)
+        {
+            var directiveText = Name;
+
             try
             {
                 // If the path is a directory like '../lib', transform it to a project file path like '../lib/lib.csproj'.
-                // Also normalize blackslashes to forward slashes to ensure the directive works on all platforms.
-                // https://github.com/dotnet/sdk/issues/51487: Behavior should not depend on process current directory
-                var sourceDirectory = Path.GetDirectoryName(context.SourceFile.Path) ?? ".";
+                // Also normalize backslashes to forward slashes to ensure the directive works on all platforms.
+                var sourceDirectory = Path.GetDirectoryName(sourceFile.Path) ?? ".";
```

Comment:
I think we should restore the tracking comment.
```suggestion
                // https://github.com/dotnet/sdk/issues/51487: Behavior should not depend on process current directory
                var sourceDirectory = Path.GetDirectoryName(sourceFile.Path) ?? ".";
```

Reply:
I don't think this code path depends on the current directory. There are even tests for that - you can run `dotnet project convert` from any working directory and it should work the same way.

Reply:
Ah, you're right, I completely missed that part, thanks!